test/ComputeDegree.lean

```diff
@@ -0,0 +1,230 @@
+example : natDegree (- C z * X ^ 5 + (monomial 2 5) ^ 2 - 0 + 1 + IntCast.intCast 1 +
+    NatCast.natCast 1 + (z : ℤ[X]) + (n : ℤ[X]) + f) ≤ 5 := by
+  compute_degree!
+
+example [Semiring R] : natDegree (OfNat.ofNat (OfNat.ofNat 0) : R[X]) ≤ 0 := by
+  compute_degree
+
+/--  Flows through all the matches in `compute_degree`, with a `degree _ ≤ _` goal. -/
+example : degree (- C z * X ^ 5 + (C 0 + monomial 2 5) ^ 2 - 0 + 1 + IntCast.intCast 1 +
+    NatCast.natCast 1 + (z : ℤ[X]) + (n : ℤ[X]) + f) ≤ 5 := by
+  set k := f with _h₀
+  compute_degree!
+
+/--  Flows through all the matches in `compute_degree`, with a `natDegree _ = _` goal. -/
+example : natDegree (- C 1 * X ^ 5 + (C 0 + monomial 2 5) ^ 2 - 0 + 1 + IntCast.intCast 1 +
+    NatCast.natCast 1 + (z : ℤ[X]) + (n : ℤ[X])) = 5 := by
+  set k := f with _h₀
+  compute_degree!
+
+/--  Flows through all the matches in `compute_degree`, with a `degree _ = _` goal. -/
+example : degree (- C 1 * X ^ 5 + (C 0 + monomial 2 5) ^ 2 - 0 + 1 + IntCast.intCast 1 +
+    NatCast.natCast 1 + (z : ℤ[X]) + (n : ℤ[X])) = 5 := by
+  set k := f with _h₀
+  compute_degree!
+
+example : degree
+    ((C 1 * X ^ 2 + C 2 * X + C 3) * (C 0 * X ^ 0 + C 2 * X ^ 1 + C 1 * X ^ 5) ^ 4) = 22 := by
+  compute_degree!
+
+--  Tests `Nat.cast_withBot`
+example [Semiring R] [Nontrivial R] {n : ℕ} : degree (X ^ n : R[X]) = n := by compute_degree!
+
+example [Nontrivial R] [Ring R] : degree
+    (1 + X + X ^ 2 - X ^ 5 - X ^ 6 - 2 * X ^ 7 - X ^ 8 - X ^ 9 + X ^ 12 + X ^ 13 + X ^ 14 +
+        X ^ 15 + X ^ 16 + X ^ 17 - X ^ 20 - X ^ 22 - X ^ 24 - X ^ 26 - X ^ 28 + X ^ 31 + X ^ 32 +
+        X ^ 33 + X ^ 34 + X ^ 35 + X ^ 36 - X ^ 39 - X ^ 40 - 2 * X ^ 41 - X ^ 42 - X ^ 43 +
+        X ^ 46 + X ^ 47 + X ^ 48 : R[X]) = 48 := by
+  compute_degree!
+
+/--  Flows through all the matches in `compute_degree`, with a `degree _ ≤ _` goal. -/
+example [Ring R] (g : R[X]) (hg : degree g ≤ 5) : degree (- C (z : R) * X ^ 5 + (monomial 2 5) ^ 2
+    - 0 + 1 + IntCast.intCast 1 + NatCast.natCast 1 + (z : R[X]) + (n : R[X]) + g) ≤ 5 := by
+  set k := g with _h₀
+  compute_degree!
+
+example {N : WithBot ℕ} (nN : n ≤ N) : degree (- C z * X ^ n) ≤ N := by compute_degree!
+
+example [Ring R] : coeff (1 : R[X]) 0 = 1 := by compute_degree!
+
+example [Ring R] : coeff (1 : R[X]) 2 = 0 := by compute_degree!
+
+example [Ring R] : coeff (1 : R[X]) n = if 0 = n then 1 else 0 := by compute_degree!
+
+example [Ring R] (h : (0 : R) = 6) : coeff (1 : R[X]) 1 = 6 := by compute_degree!
+
+/-! Test error messages -/
+
+/--
+error: The given degree is '0'.  However,
+
+* the coefficient of degree 0 may be zero
+* there may be a term of naïve degree 2
+* there is at least one term of naïve degree 2
+-/
+#guard_msgs in
+example : natDegree (X + X ^ 2 : ℕ[X]) = 0 := by compute_degree!
+
+/--
+error: 'compute_degree' inapplicable.  The goal
+
+   natDegree X ≠ 0
+
+is expected to be '≤' or '=', instead of 'Ne'.
```

Comment:
Generally tactics don't insert all these extra newlines, which is why I removed them.
```suggestion
error: 'compute_degree' inapplicable. The goal
  natDegree X ≠ 0
is expected to be '≤' or '='.
```